src/test/rsm_tests.cpp

```diff
@@ -0,0 +1,201 @@
+    rsm.unlock(std::this_thread::get_id());
+}
+
+void delta()
+{
+    // sleep to ensure epsilon got a shared lock in. we can test this
+    // by trying to share lock, should return false
+    MilliSleep(200);
+    BOOST_CHECK_EQUAL(rsm.try_lock(std::this_thread::get_id()), false);
+    // lock shared ourselves, there should be 2 shared locks at this time
+    rsm.lock_shared(std::this_thread::get_id());
+    // we sleep to ensure epsilon unlocked
+    MilliSleep(3000);
+    // try to request a lock change which should be handled internally
+    rsm.lock(std::this_thread::get_id());
+    // if this was successful we test that we have this lock by checking the
+    // size in the other thread
+    rsm_guarded_vector.push_back(0);
+    rsm_guarded_vector.push_back(1);
+    rsm_guarded_vector.push_back(2);
+    rsm_guarded_vector.push_back(3);
+    rsm.unlock(std::this_thread::get_id());
+    // we should only have a shared lock now. sleep while epsilon catches up
+    MilliSleep(3000);
+    // we should be able to unlock our shared mutex now with no errors
+    BOOST_CHECK_NO_THROW(rsm.unlock_shared(std::this_thread::get_id()));
+}
+
+void epsilon()
+{
+    rsm.lock_shared(std::this_thread::get_id());
+    // give time for dela to lock shared (67)
+    MilliSleep(5000);
+    rsm.unlock_shared(std::this_thread::get_id());
+    // sleep for 500 to give delta time to do its thing (71)
+    MilliSleep(500);
+    // try to lock shared, we should be able to
+    bool lockresult = rsm.try_lock_shared(std::this_thread::get_id());
+    BOOST_CHECK_EQUAL(lockresult, true);
+    // check size
+    BOOST_CHECK_EQUAL(rsm_guarded_vector.size(), 4);
+    // we only have one lock, but try to unlock twice.
+    rsm.unlock_shared(std::this_thread::get_id());
+    // on the second unlock we should get an assertion (std::abort)
+    BOOST_CHECK_THROW(rsm.unlock_shared(std::this_thread::get_id()), std::logic_error);
+    // at this point we should be all unlocked. before we finish the thread lets
+    // make sure delta still has it locked by trying to lock
+    BOOST_CHECK_EQUAL(rsm.try_lock(std::this_thread::get_id()), false);
+}
+
+void zeta()
+{
+    rsm.lock_shared(std::this_thread::get_id());
+    // give time for theta to lock shared, eta to lock, and theta to ask for promotion
+    MilliSleep(4000);
+    rsm.unlock_shared(std::this_thread::get_id());
+}
+
+void eta()
+{
+    rsm.lock(std::this_thread::get_id());
+    rsm_guarded_vector.push_back(4);
+    rsm.unlock(std::this_thread::get_id());
+
+}
+
+void theta()
+{
+    rsm.lock_shared(std::this_thread::get_id());
+    // give time for eta to get in line to lock exclusive
+    MilliSleep(500);
+    bool promoted = rsm.try_promotion(std::this_thread::get_id());
+    BOOST_CHECK_EQUAL(promoted, true);
+    rsm_guarded_vector.push_back(7);
+    rsm.unlock(std::this_thread::get_id());
+    rsm.unlock_shared(std::this_thread::get_id());
+}
+
+BOOST_AUTO_TEST_CASE(RsmTest)
+{
+    std::thread first (alpha);
```

Comment:
Would like to see a more commenting on what the intent of each test is and what the expected outcomes

Reply:
in an address to all 3 comments, I will reformat the tests file today for clarity about what each test is testing and to make sure I cover any missed scenarios.